### ISSUE TITLE
Charset appended to multipart parts regardless of the media type already specified

### ISSUE THREAD

**Issue by pdolezal** [quarkusio/quarkus#30262]
### Describe the bug

We use `org.jboss.resteasy.plugins.providers.multipart.MultipartFormDataInput` parameter for a JAX-RS endpoint method that consumes `multipart/form-data`. When we fetch the media type of a part, it has always `charset=UTF-8` specified, regardless of the media type that the client actually specified (so, we can get things like `application/octet-stream; charset=UTF-8`, which is absurd).

We tried to set `quarkus.resteasy-reactive.multipart.input-part.default-charset` to an empty value in `application.properties`, but this is not supported, there simply must be a valid `Charset` name, otherwise the container fails to start because of the value is not acceptable.

Anyway, we need to get the media type exactly as the client specified, so not having the option to suppress any default charset is quite a problem.

### Expected behavior

It should be possible to specify no default charset at all, so that we can get the media type as actually specified by the client. It might be acceptable if the default charset is applied for types when it makes sense (probably just `text/*`).

### Actual behavior

The endpoint receives the form data with the media type with the default charset specified always.

### How to Reproduce?

```
import javax.ws.rs.Consumes;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.core.MediaType;
import org.jboss.resteasy.plugins.providers.multipart.InputPart;
import org.jboss.resteasy.plugins.providers.multipart.MultipartFormDataInput;

@Path("/test")
public final class TestEndpoint {

    @POST
    @Consumes(MediaType.MULTIPART_FORM_DATA)
    public void test(MultipartFormDataInput form) {
        final var parts = form.getParts();

        if (parts != null) {
            parts.stream().map(InputPart::getMediaType).forEach(System.out::println);
        }
    }
}
```

Send a multipart request then, e.g.:
```
curl -X POST \
-H 'content-type: multipart/form-data; boundary=f555d9c6-f40f-4d62-a217-f64a58d75b50' \
--data-raw '--f555d9c6-f40f-4d62-a217-f64a58d75b50
Content-Disposition: form-data; name="data"
Content-Type: application/octet-stream

ABC

--f555d9c6-f40f-4d62-a217-f64a58d75b50--
' http://localhost:8080/test
```

### Quarkus version or git rev

2.15.2

**Comment by geoand** [comment]
Any chance you can attach a sample application we can use to debug and test?

**Comment by pdolezal** [comment]
Here it is. Use the `curl` above to invoke the endpoint to see the content type as the endpoint provides it to the business code.
[extra-charset.zip](https://github.com/quarkusio/quarkus/files/10380044/extra-charset.zip)

**Comment by geoand** [comment]
@jamezp this seems to be a RESTEasy specific issue. When looking through it, I traced the problem to how the charset is added in [this](https://github.com/resteasy/resteasy/blob/4.7.7.Final/providers/multipart/src/main/java/org/jboss/resteasy/plugins/providers/multipart/MultipartInputImpl.java#L156) constructor.

**Comment by geoand** [comment]
It's likely a combination. Quarkus defines a standard charset which RESTEasy uses, but seems to apply it to every media type - which is not really correct

**Comment by jamezp** [comment]
Quarkus sets the default charset in a [filter](https://github.com/quarkusio/quarkus/blob/04aa95308cc3296ec89871932cfae101daf9d46a/extensions/resteasy-classic/resteasy-multipart/runtime/src/main/java/io/quarkus/resteasy/multipart/runtime/MultipartInputPartConfigContainerRequestFilter.java#L23). I'm not too sure what RESTEasy can do about that really.

There's also not really a plan for any more 4.7 releases of RESTEasy.

**Comment by geoand** [comment]
Right, that sets a property that RESTEasy then uses in the code I mentioned earlier

**Comment by jamezp** [comment]
What would the expectation be from RESTEasy here? We can't just remove adding that as it would break backwards compatibility.

**Comment by geoand** [comment]
That's absolutely understood. From my POV, RESTEasy should only add the default charset to media types that make sense (I think we do that in RR, I'll have to find the code when I'm at my machine tomorrow).

**Comment by pdolezal** [comment]
> What would the expectation be from RESTEasy here? We can't just remove adding that as it would break backwards compatibility.

Could it be possible to make `quarkus.resteasy-reactive.multipart.input-part.default-charset` optional, so that it could be set to an empty value (and then no charset would be appended by default)?

**Comment by jamezp** [comment]
@pdolezal A workaround might be to add another filter which executes after the Quarkus one which removes the request attribute.

**Comment by jamezp** [comment]
RESTEasy only adds it because the `InputPart.DEFAULT_CHARSET_PROPERTY` is explicitly set.

**Comment by geoand** [comment]
Yes, of course. But my point is that it should probably not add it unconditionally

**Comment by jnbdz** [comment]
Still reproduces on current main (6ce0ccbb857). I traced it end to end, and there is a wrinkle that I think rules out the obvious fix, so posting the details.

**Reproduction.** Posting three parts, each with an explicit `Content-Type` and no charset, gives:

```
binary=application/octet-stream;charset=UTF-8
text=text/plain;charset=UTF-8
json=application/json;charset=UTF-8
```

**Mechanism.** As @jamezp said, Quarkus sets `InputPart.DEFAULT_CHARSET_PROPERTY` unconditionally in `MultipartInputPartConfigContainerRequestFilter`. In `MultipartInputImpl$PartImpl` the relevant sequence is:

1. parse the part's declared `Content-Type`
2. if the part has **no** charset of its own **and** `defaultPartCharset != null` → apply it to the part's *declared* type (this is what hits `application/octet-stream`)
3. otherwise, if the type is `text/*` → apply `us-ascii`

**Why just dropping the property does not work.** I tried exactly that (equivalent to @jamezp's suggested workaround of a later filter clearing the attribute). Non-text parts become correct:

```
binary=application/octet-stream
text=text/plain;charset=us-ascii
json=application/json
```

but text parts now fall into step 3 and get `us-ascii`, which silently mangles non-ASCII content. `MultipartEncodingTest` in `resteasy-multipart/deployment` fails on that — it asserts that `Text with UTF-8 accented characters: é à è` survives a round trip with the default config. So the workaround trades an absurd charset on binary parts for data corruption on text parts, and it is not a fix Quarkus can ship as-is.

**On making the config optional** (@pdolezal's suggestion): `quarkus.resteasy.multipart.input-part.default-charset` is typed `Charset` with `@WithDefault("UTF-8")`, so it cannot express "none" today. Making it `Optional<Charset>` is doable — there is no `Optional<Charset>` precedent in Quarkus, but `Optional<String> charset()` exists in the reactive MySQL config — however per the above it would give users a choice between "absurd charset on binary parts" and "`us-ascii` on text parts", which is not much of a choice.

**On the upstream side.** I looked at the RESTEasy source before suggesting anything further, and the obvious "only apply it to `text/*`" change is **not** safe: `InputPartDefaultCharsetOverwriteTest`, the regression test for RESTEASY-723, explicitly asserts that a part declared `application/xml` comes back as `application/xml; charset=US-ASCII` / `charset=UTF-8` / `charset=UTF-16`. Applying the configured default charset to non-text types is deliberate there, so gating on `text/*` would break it.

What does look unconstrained is `application/octet-stream`: those cases in that test go through `doTestByteArray`, which asserts the response **bytes** and never the content type. So a narrow change — skip the default charset only for media types where a charset is meaningless (octet-stream, and arguably other binary subtypes) — would fix the reported case without touching RESTEASY-723's expectations.

FWIW I could not find an existing RESTEasy issue tracking this.

I have opened that narrower change upstream: resteasy/resteasy#5022. `MultipartInputImpl$PartImpl` skips the default charset only for `application/octet-stream`; `text/plain` and `application/xml` keep receiving it, and the test in that PR asserts all three, so the RESTEASY-723 expectations are preserved. Without the change the octet-stream assertion fails with exactly the `application/octet-stream;charset=UTF-8` shown at the top. Once it lands in a RESTEasy release that Quarkus picks up, the reported case is fixed with no Quarkus-side change needed.

---

*Edited twice. First to correct the proposed fix: my original version suggested gating on `text/*` (plus `+json`/`+xml` suffixes), which would break the RESTEASY-723 regression test that pins `application/xml` + charset, so it was narrowed to `application/octet-stream`. Then to replace the closing question with the link to the upstream PR, resteasy/resteasy#5022. Everything else is unchanged.*
